### ISSUE TITLE
5.0.0-RC3 BUG - Class reloading stopped working with 5.0.0.RC2 - io.micronaut.ast.groovy.annotation.GroovyAnnotationMetadataBuilder does not define or inherit

### ISSUE THREAD

**Issue by codeconsole** [grails/grails-core#12065]
# Step 1 - Verify 5.0.0.RC1 Works
```
git checkout https://github.com/codeconsole/website
cd website
git checkout 5.0.0.RC1
sdk use grails 5.0.0.RC1
grails run-app
```

After Grails has finished starting up:
```
touch grails-app/domain/website/Sample.groovy
```

Expected behavior, server reloads every time touch is run or a file is updated.

```
Using grails version 5.0.0.RC1 in this shell.
| Resolving Dependencies. Please wait...

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.9/userguide/command_line_interface.html#sec:command_line_warnings

| Running application...
Grails application running at http://localhost:8080 in environment: development
File ~/Documents/workspace/website/grails-app/domain/website/Sample.groovy changed, recompiling...
Grails application running at http://localhost:8080 in environment: development
File ~/Documents/workspace/website/grails-app/domain/website/Sample.groovy changed, recompiling...
Grails application running at http://localhost:8080 in environment: development
File ~/Documents/workspace/website/grails-app/domain/website/Sample.groovy changed, recompiling...
Grails application running at http://localhost:8080 in environment: development
<===========--> 85% EXECUTING [1m 13s]
> :bootRun
```

# Step 2 - Verify 5.0.0-RC3 is Broken
```
git checkout https://github.com/codeconsole/website
cd website
git checkout 5.0.0-RC3
sdk use grails 5.0.0-3RC3
grails run-app
```

After Grails has finished starting up:
```
touch grails-app/domain/website/Sample.groovy
```

Expected behavior, server reloads every time touch is run or a file is updated, but what happens is change is detected but no restart happens.

```
Using grails version 5.0.0-RC3 in this shell.
| Resolving Dependencies. Please wait...

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.9.1/userguide/command_line_interface.html#sec:command_line_warnings

| Running application...
Grails application running at http://localhost:8080 in environment: development
File ~/Documents/workspace/website/grails-app/domain/website/Sample.groovy changed, recompiling...
<===========--> 85% EXECUTING [38s]
> :bootRun
```

# Step 3 - Optional - Open 5.0.0-RC3 version in IntelliJ and notice that restart ends in exception which could be the underlying problem:

```
File ~/Documents/workspace/pixoto3/grails-app/services/com/pixoto/image/ImageService.groovy changed, recompiling...
Exception in thread "Thread-13" java.lang.AbstractMethodError: Receiver class io.micronaut.ast.groovy.annotation.GroovyAnnotationMetadataBuilder does not define or inherit an implementation of the resolved method 'abstract boolean isInheritedAnnotation(java.lang.Object)' of abstract class io.micronaut.inject.annotation.AbstractAnnotationMetadataBuilder.
	at io.micronaut.inject.annotation.AbstractAnnotationMetadataBuilder.processAnnotationStereotype(AbstractAnnotationMetadataBuilder.java:1511)
	at io.micronaut.inject.annotation.AbstractAnnotationMetadataBuilder.includeAnnotations(AbstractAnnotationMetadataBuilder.java:1176)
	at io.micronaut.inject.annotation.AbstractAnnotationMetadataBuilder.buildInternal(AbstractAnnotationMetadataBuilder.java:1097)
	at io.micronaut.inject.annotation.AbstractAnnotationMetadataBuilder.build(AbstractAnnotationMetadataBuilder.java:245)
	at io.micronaut.inject.annotation.AbstractAnnotationMetadataBuilder.build(AbstractAnnotationMetadataBuilder.java:226)
	at io.micronaut.ast.groovy.utils.AstAnnotationUtils$_getAnnotationMetadata_closure1.doCall(AstAnnotationUtils.groovy:54)
	at jdk.internal.reflect.GeneratedMethodAccessor21.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:263)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at groovy.lang.Closure.call(Closure.java:412)
	at org.codehaus.groovy.runtime.ConvertedClosure.invokeCustom(ConvertedClosure.java:50)
	at org.codehaus.groovy.runtime.ConversionHandler.invoke(ConversionHandler.java:110)
	at com.sun.proxy.$Proxy19.apply(Unknown Source)
	at io.micronaut.core.util.clhm.ConcurrentLinkedHashMap.lambda$compute$0(ConcurrentLinkedHashMap.java:721)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
	at io.micronaut.core.util.clhm.ConcurrentLinkedHashMap.compute(ConcurrentLinkedHashMap.java:733)
	at io.micronaut.core.util.clhm.ConcurrentLinkedHashMap.computeIfAbsent(ConcurrentLinkedHashMap.java:710)
	at io.micronaut.ast.groovy.utils.AstAnnotationUtils.getAnnotationMetadata(AstAnnotationUtils.groovy:53)
	at io.micronaut.ast.groovy.visitor.LoadedVisitor.matches(LoadedVisitor.groovy:114)
	at io.micronaut.ast.groovy.TypeElementVisitorTransform$_visit_closure1.doCall(TypeElementVisitorTransform.groovy:81)
	at jdk.internal.reflect.GeneratedMethodAccessor225.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:263)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:38)
	at org.codehaus.groovy.runtime.callsite.BooleanReturningMethodInvoker.invoke(BooleanReturningMethodInvoker.java:49)
	at org.codehaus.groovy.runtime.callsite.BooleanClosureWrapper.call(BooleanClosureWrapper.java:52)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.findAll(DefaultGroovyMethods.java:4971)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.findAll(DefaultGroovyMethods.java:4843)
	at io.micronaut.ast.groovy.TypeElementVisitorTransform.visit(TypeElementVisitorTransform.groovy:81)
	at org.codehaus.groovy.transform.ASTTransformationVisitor.lambda$addPhaseOperationsForGlobalTransforms$4(ASTTransformationVisitor.java:337)
	at org.codehaus.groovy.control.CompilationUnit$ISourceUnitOperation.doPhaseOperation(CompilationUnit.java:901)
	at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:671)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:635)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:610)
	at grails.boot.GrailsApp.compileGroovyFile(GrailsApp.groovy:410)
	at grails.boot.GrailsApp.recompile(GrailsApp.groovy:396)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:351)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.callCurrent(PogoMetaClassSite.java:61)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:171)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:203)
	at grails.boot.GrailsApp$_enableDevelopmentModeWatch_closure3.doCall(GrailsApp.groovy:339)
	at grails.boot.GrailsApp$_enableDevelopmentModeWatch_closure3.doCall(GrailsApp.groovy)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:263)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at groovy.lang.Closure.call(Closure.java:412)
	at groovy.lang.Closure.call(Closure.java:406)
	at groovy.lang.Closure.run(Closure.java:493)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by puneetbehl** [comment]
It seems like somehow the runtimeClasspath is pointing to micronaut-inject-groovy version 2.2.1 which might be the problem.

**Comment by puneetbehl** [comment]
Forcing the version using following will fix the problem:
```
configurations.all {
    resolutionStrategy.eachDependency {
        if (it.requested.group == 'io.micronaut' && it.requested.name == 'micronaut-inject-groovy') {
            it.useVersion('3.0.1')
        }
    }
}
```

**Comment by puneetbehl** [comment]
Updating `grailsGradlePluginVersion=5.0.0-RC3` in gradle.properties file should fix the problem.

**Comment by codeconsole** [comment]
@puneetbehl Did 5.0.0-RC3 get published?
`Could not find org.grails:grails-gradle-plugin:5.0.0-RC3.`

https://repo.grails.org/ui/native/core/org/grails/grails-gradle-plugin/5.0.0-RC3/grails-gradle-plugin-5.0.0-RC3.pom

https://repo.grails.org/ui/native/core/org/grails/grails-gradle-plugin/

I can confirm that 
```
grailsGradlePluginVersion=5.0.0-SNAPSHOT
```

works. Thank you.

**Comment by puneetbehl** [comment]
Fixed with 6a78a6709bd49aa3b9050689126e1a1cf2a4691f
